Bluetooth: host: Support fragmented L2CAP header

thomasstenersen opened the pull request:
Fixes: #26900.

A controller may fragment an L2CAP SDUs in any way it sees fit,
including fragmenting the L2CAP header. Likewise, the receiving
controller may send the fragmented header as ACL data to the host.

The Zephyr host assumed that a `BT_ACL_START` was at least 2 bytes long,
and consequently read the two-byte length field from the buffer without
length checks.

This commit allows the `BT_ACL_START` to be less than two bytes,
updating the `conn->rx_len` onces the `BT_ACL_CONT` with the remaining
part of the length field has been received.

Signed-off-by: Thomas Stenersen <thomas.stenersen@nordicsemi.no>

Commit:
Bluetooth: host: Support fragmented L2CAP header

Fixes issue #26900.

A controller may fragment an L2CAP SDUs in any way it sees fit,
including fragmenting the L2CAP header. Likewise, the receiving
controller may send the fragmented header as ACL data to the host.

The Zephyr host assumed that a `BT_ACL_START` was at least 2 bytes long,
and consequently read the two-byte length field from the buffer without
length checks.

This commit allows the `BT_ACL_START` to be less than two bytes,
updating the `conn->rx_len` onces the `BT_ACL_CONT` with the remaining
part of the length field has been received.

Signed-off-by: Thomas Stenersen <thomas.stenersen@nordicsemi.no>

thomasstenersen commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1204,14 +1204,36 @@ void bt_conn_recv(struct bt_conn *conn, struct net_buf *buf, uint8_t flags)
+			bt_conn_reset_rx_state(conn);
+		}
+
+		if (!buf->len) {
```
I'm a bit unsure whether it is allowed or not for the controller to send zero-length ACL data to the host. The other way is definitely allowed (and is tested in LL conformance tests). In any case, the host shouldn't misbehave if it does receive zero-length ACL data.

Vudentz replied:
I guess 0 length is also possible, at least I don't see anything on the spec preventing that.

Vudentz commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1223,13 +1245,33 @@ void bt_conn_recv(struct bt_conn *conn, struct net_buf *buf, uint8_t flags)
+		 * yet?
+		 */
+		if (conn->rx->len < sizeof(uint16_t)) {
+			uint16_t len = (conn->rx->data[0] | buf->data[0] << 8);
```
Not following this one, we will need to keep appending data until there are at least 2 bytes since the start could have 0 bytes there could be necessary to have 2+ rounds to read the entire header if the controller send them one byte at time.

thomasstenersen replied:
This logic is a bit convoluted, I agree. The possible scenarios are:

- 0-byte START => drop buffer on L1214.
- 1-byte START, 0-byte CONT => drop continuation fragment on L1257.
- 1-byte START, >=1-byte CONT => calculate the `conn->rx_len` as done here
- >1-byte START => Normal case

Any suggestions on how to make it clearer? We can at least move the check for empty buf outside the switch.

Vudentz replied:
If we drop the buffer at L1214 that means the conn->rx won't be initialized so CONT will cause the Unexpected L2CAP continuation which I guess that needs fixing if we really intend to make this generic. I was actually thinking if it wouldn't be better to reassemble first, problem is there is no flag to indicate end of fragments like there is for ISO which means we need to reconstruct the length while parsing the fragments, perhaps we need to having a separate helper function that does the appending would be better here which we keep calling until there is enough data to set the rx_len.

thomasstenersen replied:
Right, I see. I'll see if I can clean the logic up a bit.

thomasstenersen replied:
I added a suggested re-write in https://github.com/zephyrproject-rtos/zephyr/pull/27169/commits/145257454d3bcd17b206b182878aabebda5e6cc3. It should clean up the logic quite a bit.

thomasstenersen replied:
Did you have a look @Vudentz ? I removed the `[wip]` tag from the commit to let CI run, the commits can be squashed later.
https://github.com/zephyrproject-rtos/zephyr/pull/27169/commits/4adcaafe0b1ccb0bfc3f0968c47c0b7b8e0c9c01

Commit:
Bluetooth: host: Refactor bt_conn_recv

Re-structures the `bt_conn_recv()` to support more generic appending of
new ACL data. This also handles the possibility of receiving 0-length
ACL_START and ACL_CONT fragments, whilst maintaining existing
error-checking functionality.

Signed-off-by: Thomas Stenersen <thomas.stenersen@nordicsemi.no>

thomasstenersen commented:
@Vudentz will you get a chance to look at this one some time soon?

Vudentz commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1006,17 +942,33 @@ void bt_conn_recv(struct bt_conn *conn, struct net_buf *buf, uint8_t flags)
-	if (sizeof(*hdr) + len != buf->len) {
-		BT_ERR("ACL len mismatch (%u != %u)", len, buf->len);
-		net_buf_unref(buf);
+	const struct bt_l2cap_hdr *hdr = (struct bt_l2cap_hdr *)conn->rx->data;
```
Not a big fan of variable declaration in the middle of the function scope, also note that bt_l2cap_hdr contains the cid as well not just the len, so Id would suggest to just skip parsing the header and make bt_l2cap_recv do that using the whole buffer and if the header was not received (lenght <= 1) we should just drop.

thomasstenersen replied:
Thanks for taking the time to do another review round 👍 

> note that bt_l2cap_hdr contains the cid as well not just the len, so Id would suggest to just skip parsing the header and make bt_l2cap_recv do that using the whole buffer and if the header was not received (lenght <= 1) we should just drop.

I don't understand what you're asking me to do here. I'm not touching the `cid`, only using the `struct bt_l2cap_hdr` for retrieving the length field and for its size when calculating the total length of the ACL data. We need to parse the length field to know when to propagate the buffer to L2CAP, right? You're not saying you want to split this responsibility of reassembling across here and L2CAP? If you don't like the fact that the `struct bt_l2cap_hdr` type is used, it is possible to do something like

```c
const uint16_t l2cap_len = sys_get_le16(buf->data);
const uint16_t acl_total_len = l2cap_len + sizeof(struct bt_l2cap_hdr);
```
Regarding 
> Not a big fan of variable declaration in the middle of the function scope

Note that we cannot retrieve this length until we know that `buf->len >= 2`, which is not the case at the beginning of the function. Using C89-style would then force us to drop the `const` qualifier to declare a variable that's only used 60+ lines below. If we do strictly follow C89-style, then that's fine, I can declare the variable at the top of the function.

Vudentz commented on subsys/bluetooth/host/conn.c:
```diff
@@ -903,40 +899,13 @@ void bt_conn_recv(struct bt_conn *conn, struct net_buf *buf, uint8_t flags)
+		       (buf->len < sizeof(uint16_t)) ?
+		       0 :
+		       sys_le16_to_cpu(
+			       ((struct bt_l2cap_hdr *)buf->data)->len));
```
Actually it is just the length that should be read, not bt_l2cap_hdr, so this should probably be something like:

buf->len < sizeof(uint16_t) ? buf->len : sys_le16_to_cpu(buf->data);

thomasstenersen replied:
I originally did this, but changed it to make it clearer that it's the l2cap header length field we're retrieving. No problem switching it back though.

Commit:
Bluetooth: host: Fetch L2CAP length directly

This commit will be squashed after review.

Signed-off-by: Thomas Stenersen <thomas.stenersen@nordicsemi.no>

Commit:
Bluetooth: host: Use C89-style

To be squashed after review.

Signed-off-by: Thomas Stenersen <thomas.stenersen@nordicsemi.no>

joerchan commented on subsys/bluetooth/host/conn.c:
```diff
@@ -964,17 +941,32 @@ void bt_conn_recv(struct bt_conn *conn, struct net_buf *buf, uint8_t flags)
-	if (sizeof(*hdr) + len != buf->len) {
-		BT_ERR("ACL len mismatch (%u != %u)", len, buf->len);
-		net_buf_unref(buf);
+	acl_total_len = sys_get_le16(buf->data) + sizeof(struct bt_l2cap_hdr);
```
buf is unref'ed, and length field could be split.
```suggestion
	acl_total_len = sys_get_le16(conn->rx->data) + sizeof(struct bt_l2cap_hdr);
```

Commit:
Bluetooh: Host: Fix wrong buffer in acl_total_len calc

To be squashed.

Co-authored-by: Joakim Andersson <joakim.andersson@nordicsemi.no>
Signed-off-by: Thomas Stenersen <thomas.stenersen@nordicsemi.no>

Vudentz reviewed (approved):
LGTM

joerchan commented:
@Vudentz Has BT_ISO the same problem with a fragmented header?
